README.md
# Jellyfish [![Pipeline status](https://gitlab.com/godfat/jellyfish/badges/master/pipeline.svg)](https://gitlab.com/godfat/jellyfish/-/pipelines)

by Lin Jen-Shin ([godfat](https://godfat.org))

![logo](https://github.com/godfat/jellyfish/raw/master/jellyfish.png)

## LINKS:

* [github](https://github.com/godfat/jellyfish)
* [rubygems](https://rubygems.org/gems/jellyfish)
* [rdoc](https://rubydoc.info/github/godfat/jellyfish)
* [issues](https://github.com/godfat/jellyfish/issues) (feel free to ask for support)

## DESCRIPTION:

Pico web framework for building API-centric web applications.
For Rack applications or Rack middleware. Around 250 lines of code.

Check [jellyfish-contrib][] for extra extensions.

[jellyfish-contrib]: https://github.com/godfat/jellyfish-contrib

## DESIGN:

* Learn the HTTP way instead of using some pointless helpers.
* Learn the Rack way instead of wrapping around Rack functionalities.
* Learn regular expression for routes instead of custom syntax.
* Embrace simplicity over convenience.
* Don't make things complicated only for _some_ convenience, but for
  _great_ convenience, or simply stay simple for simplicity.
* More features are added as extensions.
* Consider use [rack-protection][] if you're not only building an API server.
* Consider use [websocket_parser][] if you're trying to use WebSocket.
  Please check example below.

[rack-protection]: https://github.com/rkh/rack-protection
[websocket_parser]: https://github.com/afcapel/websocket_parser

## FEATURES:

* Minimal
* Simple
* Modular
* No templates (You could use [tilt](https://github.com/rtomayko/tilt))
* No ORM (You could use [sequel](https://sequel.jeremyevans.net))
* No `dup` in `call`
* Regular expression routes, e.g. `get %r{^/(?<id>\d+)$}`
* String routes, e.g. `get '/'`
* Custom routes, e.g. `get Matcher.new`
* Build for either Rack applications or Rack middleware
* Include extensions for more features (checkout [jellyfish-contrib][])

## WHY?

Because Sinatra is too complex and inconsistent for me.

## REQUIREMENTS:

* Tested with MRI (official CRuby) and JRuby.

## INSTALLATION:

    gem install jellyfish

## SYNOPSIS:

You could also take a look at [config.ru](config.ru) as an example.

### Hello Jellyfish, your lovely config.ru

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  get '/' do
    "Jelly Kelly\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /
[200,
 {'content-length' => '12', 'content-type' => 'text/plain'},
 ["Jelly Kelly\n"]]
-->

### Regular expression routes

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  get %r{^/(?<id>\d+)$} do |match|
    "Jelly ##{match[:id]}\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /123
[200,
 {'content-length' => '11', 'content-type' => 'text/plain'},
 ["Jelly #123\n"]]
-->

### Custom matcher routes

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  class Matcher
    def match path
      path.reverse == 'match/'
    end
  end
  get Matcher.new do |match|
    "#{match}\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /hctam
[200,
 {'content-length' => '5', 'content-type' => 'text/plain'},
 ["true\n"]]
-->

### Different HTTP status and custom headers

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  post '/' do
    headers       'X-Jellyfish-Life' => '100'
    headers_merge 'X-Jellyfish-Mana' => '200'
    body "Jellyfish 100/200\n"
    status 201
    'return is ignored if body has already been set'
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
POST /
[201,
 {'content-length' => '18', 'content-type' => 'text/plain',
  'X-Jellyfish-Life' => '100', 'X-Jellyfish-Mana' => '200'},
 ["Jellyfish 100/200\n"]]
-->

### Redirect helper

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  get '/lookup' do
    found "#{env['rack.url_scheme']}://#{env['HTTP_HOST']}/"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /lookup host
body = File.read("#{File.dirname(
  File.expand_path(__FILE__))}/../lib/jellyfish/public/302.html").
  gsub('VAR_URL', 'http://host/')
[302,
 {'content-length' => body.bytesize.to_s, 'content-type' => 'text/html',
  'location' => 'http://host/'},
 [body]]
-->

### Crash-proof

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  get '/crash' do
    raise 'crash'
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /crash
body = File.read("#{File.dirname(
  File.expand_path(__FILE__))}/../lib/jellyfish/public/500.html")
[500,
 {'content-length' => body.bytesize.to_s, 'content-type' => 'text/html'},
 [body]]
-->

### Custom error handler

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  handle NameError do |e|
    status 403
    "No one hears you: #{e.backtrace.first}\n"
  end
  get '/yell' do
    yell
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /yell
body = case RUBY_ENGINE
       when 'jruby'
         "No one hears you: (eval):9:in `block in Tank'\n"
       when 'rbx'
         "No one hears you: core/zed.rb:1370:in `yell (method_missing)'\n"
       else
         "No one hears you: (eval):9:in `block in <class:Tank>'\n"
       end
[403,
 {'content-length' => body.bytesize.to_s, 'content-type' => 'text/plain'},
 [body]]
-->

### Custom error 404 handler

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  handle Jellyfish::NotFound do |e|
    status 404
    "You found nothing."
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /
[404,
 {'content-length' => '18', 'content-type' => 'text/plain'},
 ["You found nothing."]]
-->

### Custom error handler for multiple errors

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  handle Jellyfish::NotFound, NameError do |e|
    status 404
    "You found nothing."
  end
  get '/yell' do
    yell
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /
[404,
 {'content-length' => '18', 'content-type' => 'text/plain'},
 ["You found nothing."]]
-->

### Access Rack::Request and params

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  get '/report' do
    "Your name is #{request.params['name']}\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /report?name=godfat
[200,
 {'content-length' => '20', 'content-type' => 'text/plain'},
 ["Your name is godfat\n"]]
-->

### Re-dispatch the request with modified env

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  get '/report' do
    status, headers, body = jellyfish.call(env.merge('PATH_INFO' => '/info'))
    self.status  status
    self.headers headers
    self.body    body
  end
  get('/info'){ "OK\n" }
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /report
[200,
 {'content-length' => '3', 'content-type' => 'text/plain'},
 ["OK\n"]]
-->

### Include custom helper in built-in controller

Basically it's the same as defining a custom controller and then
include the helper. This is merely a short hand. See next section
for defining a custom controller.

``` ruby
require 'jellyfish'
class Heater
  include Jellyfish
  get '/status' do
    temperature
  end

  module Helper
    def temperature
      "30\u{2103}\n"
    end
  end
  controller_include Helper
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Heater.new
```

<!---
GET /status
[200,
 {'content-length' => '6', 'content-type' => 'text/plain'},
 ["30\u{2103}\n"]]
-->

### Define custom controller manually

This is effectively the same as defining a helper module as above and
include it, but more flexible and extensible.

``` ruby
require 'jellyfish'
class Heater
  include Jellyfish
  get '/status' do
    temperature
  end

  class Controller < Jellyfish::Controller
    def temperature
      "30\u{2103}\n"
    end
  end
  controller Controller
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Heater.new
```

<!---
GET /status
[200,
 {'content-length' => '6', 'content-type' => 'text/plain'},
 ["30\u{2103}\n"]]
-->

### Override dispatch for processing before action

We don't have before action built-in, but we could override `dispatch` in
the controller to do the same thing. CAVEAT: Remember to call `super`.

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  controller_include Module.new{
    def dispatch
      @state = 'jumps'
      super
    end
  }

  get do
    "Jelly #{@state}.\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /123
[200,
 {'content-length' => '13', 'content-type' => 'text/plain'},
 ["Jelly jumps.\n"]]
-->

### Extension: Jellyfish::Builder, a faster Rack::Builder and Rack::URLMap

Default `Rack::Builder` and `Rack::URLMap` is routing via linear search,
which could be very slow with a large number of routes. We could use
`Jellyfish::Builder` in this case because it would compile the routes
into a regular expression, it would be matching much faster than
linear search.

Note that `Jellyfish::Builder` is not a complete compatible implementation.
The followings are intentional:

* There's no `Jellyfish::Builder.call` because it doesn't make sense in my
  opinion. Always use `Jellyfish::Builder.app` instead.

* There's no `Jellyfish::Builder.parse_file` and
  `Jellyfish::Builder.new_from_string` because Rack servers are not
  going to use `Jellyfish::Builder` to parse `config.ru` at this point.
  We could provide this if there's a need.

* `Jellyfish::URLMap` does not modify `env`, and it would call the app with
  another instance of Hash. Mutating data is a bad idea.

* All other tests passed the same test suites for `Rack::Builder` and
  `Jellyfish::URLMap`.

``` ruby
require 'jellyfish'

run Jellyfish::Builder.app{
  map '/a'   do; run lambda{ |_| [200, {}, ["a\n"]  ] }; end
  map '/b'   do; run lambda{ |_| [200, {}, ["b\n"]  ] }; end
  map '/c'   do; run lambda{ |_| [200, {}, ["c\n"]  ] }; end
  map '/d'   do; run lambda{ |_| [200, {}, ["d\n"]  ] }; end
  map '/e' do
    map '/f' do; run lambda{ |_| [200, {}, ["e/f\n"]] }; end
    map '/g' do; run lambda{ |_| [200, {}, ["e/g\n"]] }; end
    map '/h' do; run lambda{ |_| [200, {}, ["e/h\n"]] }; end
    map '/i' do; run lambda{ |_| [200, {}, ["e/i\n"]] }; end
    map '/'  do; run lambda{ |_| [200, {}, ["e\n"]]   }; end
  end
  map '/j'   do; run lambda{ |_| [200, {}, ["j\n"]  ] }; end
  map '/k'   do; run lambda{ |_| [200, {}, ["k\n"]  ] }; end
  map '/l'   do; run lambda{ |_| [200, {}, ["l\n"]  ] }; end
  map '/m' do
    map '/g' do; run lambda{ |_| [200, {}, ["m/g\n"]] }; end
    run lambda{ |_| [200, {}, ["m\n"]  ] }
  end

  use Rack::ContentLength
  run lambda{ |_| [200, {}, ["/\n"]] }
}
```

<!---
GET /a
[200, {}, ["a\n"]]

GET /a/x
[200, {}, ["a\n"]]

GET /b
[200, {}, ["b\n"]]

GET /c
[200, {}, ["c\n"]]

GET /d
[200, {}, ["d\n"]]

GET /e/f
[200, {}, ["e/f\n"]]

GET /e/g
[200, {}, ["e/g\n"]]

GET /e/h
[200, {}, ["e/h\n"]]

GET /e/i
[200, {}, ["e/i\n"]]

GET /e/
[200, {}, ["e\n"]]

GET /e
[200, {}, ["e\n"]]

GET /e/x
[200, {}, ["e\n"]]

GET /j
[200, {}, ["j\n"]]

GET /k
[200, {}, ["k\n"]]

GET /l
[200, {}, ["l\n"]]

GET /m/g
[200, {}, ["m/g\n"]]

GET /m
[200, {}, ["m\n"]]

GET /m/
[200, {}, ["m\n"]]

GET /m/x
[200, {}, ["m\n"]]

GET /
[200, {'content-length' => '2'}, ["/\n"]]

GET /x
[200, {'content-length' => '2'}, ["/\n"]]

GET /ab
[200, {'content-length' => '2'}, ["/\n"]]
-->

You could try a stupid benchmark yourself:

    ruby -Ilib bench/bench_builder.rb

For a 1000 routes app, here's my result:

```
Calculating -------------------------------------
   Jellyfish::URLMap     5.726k i/100ms
        Rack::URLMap   167.000  i/100ms
-------------------------------------------------
   Jellyfish::URLMap     62.397k (± 1.2%) i/s -    314.930k
        Rack::URLMap      1.702k (± 1.5%) i/s -      8.517k

Comparison:
   Jellyfish::URLMap:    62397.3 i/s
        Rack::URLMap:     1702.0 i/s - 36.66x slower
```

#### Extension: Jellyfish::Builder#listen

`listen` is a convenient way to define routing based on the host. We could
also use `map` inside `listen` block. Here's a quick example that specifically
listen on a particular host for long-polling and all other hosts would go to
the default app.

``` ruby
require 'jellyfish'

long_poll = lambda{ |env| [200, {}, ["long_poll #{env['HTTP_HOST']}\n"]] }
fast_app  = lambda{ |env| [200, {}, ["fast_app  #{env['HTTP_HOST']}\n"]] }

run Jellyfish::Builder.app{
  listen 'slow-app' do
    run long_poll
  end

  run fast_app
}
```

<!---
GET / slow-app
[200, {}, ["long_poll slow-app\n"]]

GET / fast-app
[200, {}, ["fast_app  fast-app\n"]]
-->

##### Extension: Jellyfish::Builder#listen (`map path, host:`)

Alternatively, we could pass `host` as an argument to `map` so that the
endpoint would only listen on a specific host.

``` ruby
require 'jellyfish'

long_poll = lambda{ |env| [200, {}, ["long_poll #{env['HTTP_HOST']}\n"]] }
fast_app  = lambda{ |env| [200, {}, ["fast_app  #{env['HTTP_HOST']}\n"]] }

run Jellyfish::Builder.app{
  map '/', host: 'slow-app' do
    run long_poll
  end

  run fast_app
}
```

<!---
GET / slow-app
[200, {}, ["long_poll slow-app\n"]]

GET / fast-app
[200, {}, ["fast_app  fast-app\n"]]
-->

<!---
GET / slow-app
[200, {}, ["long_poll slow-app\n"]]

GET / fast-app
[200, {}, ["fast_app  fast-app\n"]]
-->

##### Extension: Jellyfish::Builder#listen (`map "http://#{path}"`)

Or if you really prefer the `Rack::URLMap` compatible way, then you could
just add `http://host` to your path prefix. `https` works, too.

``` ruby
require 'jellyfish'

long_poll = lambda{ |env| [200, {}, ["long_poll #{env['HTTP_HOST']}\n"]] }
fast_app  = lambda{ |env| [200, {}, ["fast_app  #{env['HTTP_HOST']}\n"]] }

run Jellyfish::Builder.app{
  map 'http://slow-app' do
    run long_poll
  end

  run fast_app
}
```

<!---
GET / slow-app
[200, {}, ["long_poll slow-app\n"]]

GET / fast-app
[200, {}, ["fast_app  fast-app\n"]]
-->

#### Extension: Jellyfish::Rewrite

`Jellyfish::Builder` is mostly compatible with `Rack::Builder`, and
`Jellyfish::Rewrite` is an extension to `Rack::Builder` which allows
you to rewrite `env['PATH_INFO']` in an easy way. In an ideal world,
we don't really need this. But in real world, we might want to have some
backward compatible API which continues to work even if the API endpoint
has already been changed.

Suppose the old API was: `/users/me`, and we want to change to `/profiles/me`,
while leaving the `/users/list` as before. We may have this:

``` ruby
require 'jellyfish'

users_api    = lambda{ |env| [200, {}, ["/users#{env['PATH_INFO']}\n"]] }
profiles_api = lambda{ |env| [200, {}, ["/profiles#{env['PATH_INFO']}\n"]] }

run Jellyfish::Builder.app{
  rewrite '/users/me' => '/me' do
    run profiles_api
  end
  map '/profiles' do
    run profiles_api
  end
  map '/users' do
    run users_api
  end
}
```

<!---
GET /users/me
[200, {}, ["/profiles/me\n"]]

GET /users/list
[200, {}, ["/users/list\n"]]

GET /profiles/me
[200, {}, ["/profiles/me\n"]]

GET /profiles/list
[200, {}, ["/profiles/list\n"]]
-->

This way, we would rewrite `/users/me` to `/profiles/me` and serve it with
our profiles API app, while leaving all other paths begin with `/users`
continue to work with the old users API app.

##### Extension: Jellyfish::Rewrite (`map path, to:`)

Note that you could also use `map path, :to` if you prefer this API more:

``` ruby
require 'jellyfish'

users_api    = lambda{ |env| [200, {}, ["/users#{env['PATH_INFO']}\n"]] }
profiles_api = lambda{ |env| [200, {}, ["/profiles#{env['PATH_INFO']}\n"]] }

run Jellyfish::Builder.app{
  map '/users/me', to: '/me' do
    run profiles_api
  end
  map '/profiles' do
    run profiles_api
  end
  map '/users' do
    run users_api
  end
}
```

<!---
GET /users/me
[200, {}, ["/profiles/me\n"]]

GET /users/list
[200, {}, ["/users/list\n"]]

GET /profiles/me
[200, {}, ["/profiles/me\n"]]

GET /profiles/list
[200, {}, ["/profiles/list\n"]]
-->

##### Extension: Jellyfish::Rewrite (`rewrite rules`)

Note that `rewrite` takes a hash which could contain more than one rule:

``` ruby
require 'jellyfish'

profiles_api = lambda{ |env| [200, {}, ["/profiles#{env['PATH_INFO']}\n"]] }

run Jellyfish::Builder.app{
  rewrite '/users/me' => '/me',
          '/users/fa' => '/fa' do
    run profiles_api
  end
}
```

<!---
GET /users/me
[200, {}, ["/profiles/me\n"]]

GET /users/fa
[200, {}, ["/profiles/fa\n"]]
-->

### Extension: NormalizedParams (with force_encoding)

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  controller_include Jellyfish::NormalizedParams

  get %r{^/(?<id>\d+)$} do
    "Jelly ##{params[:id]}\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /123
[200,
 {'content-length' => '11', 'content-type' => 'text/plain'},
 ["Jelly #123\n"]]
-->

### Extension: NormalizedPath (with unescaping)

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  controller_include Jellyfish::NormalizedPath

  get "/\u{56e7}" do
    "#{env['PATH_INFO']}=#{path_info}\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /%E5%9B%A7
[200,
 {'content-length' => '16', 'content-type' => 'text/plain'},
 ["/%E5%9B%A7=/\u{56e7}\n"]]
-->

### Extension: Using multiple extensions with custom controller

Note that the controller should be assigned lastly in order to include
modules remembered in controller_include.

``` ruby
require 'jellyfish'
class Tank
  include Jellyfish
  class MyController < Jellyfish::Controller
    include Jellyfish::WebSocket
  end
  controller_include NormalizedParams, NormalizedPath
  controller MyController

  get %r{^/(?<id>\d+)$} do
    "Jelly ##{params[:id]} jumps.\n"
  end
end
use Rack::ContentLength
use Rack::ContentType, 'text/plain'
run Tank.new
```

<!---
GET /123
[200,
 {'content-length' => '18', 'content-type' => 'text/plain'},
 ["Jelly #123 jumps.\n"]]
-->

### Jellyfish as a middleware

If the Jellyfish middleware cannot find a corresponding action, it would
then forward the request to the lower application. We call this `cascade`.

``` ruby
require 'jellyfish'
class Heater
  include Jellyfish
  get '/status' do
    "30\u{2103}\n"
  end
end

class Tank
  include Jellyfish
  get '/' do
    "Jelly Kelly\n"
  end
end

use Rack::ContentLength
use Rack::ContentType, 'text/plain'
use Heater
run Tank.new
```

<!---
GET /
[200,
 {'content-length' => '12', 'content-type' => 'text/plain'},
 ["Jelly Kelly\n"]]
-->

### Modify response as a middleware

We could also explicitly call the lower app. This would give us more
flexibility than simply forwarding it.

``` ruby
require 'jellyfish'
class Heater
  include Jellyfish
  get '/status' do
    status, headers, body = jellyfish.app.call(env)
    self.status  status
    self.headers headers
    self.body    body
    headers_merge('X-Temperature' => "30\u{2103}")
  end

CHANGES.md
# CHANGES

## Jellyfish 1.4.0 -- 2023-02-22

### Incompatible changes

* Adopted Rack 3. Technically, just lower the cases for headers.
* Internal strings are all frozen now.
* `log` and `log_error` now takes the `env` for the second argument,
  rather than the error stream.
* `handle` now takes the `env` for the third argument,
  rather than the error stream.

### Enhancements

* The default error logging will now also show `env['PATH_INFO']`.

## Jellyfish 1.3.1 -- 2018-11-11

### Bugs fixed

* Fixed `Jellyfish::Rewrite` for SCRIPT_NAME when host is also used.

## Jellyfish 1.3.0 -- 2018-11-11

### Incompatible changes

* Interface for `Builder.app`, and `Builder#to_app`, and `Rewrite.new`
  slightly changed due to fixing the bug in `Rewrite`. You shouldn't use
  those directly though.
* Now all strings allocated from Jellyfish are frozen.

### Bugs fixed

* Fixed `Jellyfish::Rewrite`. Should properly handle SCRIPT_NAME.

## Jellyfish 1.2.2 -- 2018-09-23

### Bugs fixed

* Fixed `Jellyfish::Rewrite`. Should put rewritten path in the front.

## Jellyfish 1.2.1 -- 2018-07-21

### Bugs fixed

* Fixed mapping with host in some cases (longest match goes first)
* Fixed scheme matching with https. Now it won't try to map against https
  because it's not that easy to implement and this is how `Rack::URLMap`
  works anyway.

## Jellyfish 1.2.0 -- 2018-07-14

### Incompatible changes

* `Jellyfish::NewRelic` is extracted to
  [jellyfish-contrib](https://github.com/godfat/jellyfish-contrib)

### Bugs fixed

* `Jellyfish::URLMap` would now properly handle nested maps like
  `Rack::URLMap`.

### Enhancements

* `Jellyfish::URLMap` now supports listening on a specific host like
  `Rack::URLMap`. Beside prefixing `http://`, it also supports passing
  `host: host` argument to `map`, and a `listen` directive which takes a
  block.
* Instead of using `module_eval`, DSL is now defined via `define_method` to
  make debugging easier. Performance shouldn't get hit.

## Jellyfish 1.1.1 -- 2015-12-22

### Enhancements

* Added `Jellyfish::Rewrite` as an extension to `Jellyfish::Builder`.
  Please check README.md for more detail.

## Jellyfish 1.1.0 -- 2015-09-25

### Incompatible changes

* `Jellyfish::Sinatra`, `Jellyfish::MultiActions`, and `Jellyfish::Swagger`
  were extracted to
  [jellyfish-contrib](https://github.com/godfat/jellyfish-contrib)

### Other enhancements

* Added `Jellyfish::Builder` and `Jellyfish::URLMap` which is 36 times faster
  than `Rack::Builder` and `Rack::URLMap` given an application with
  1000 routes.

## Jellyfish 1.0.2 -- 2014-12-09

* `Jellyfish::NewRelic` is fixed. Thanks Jason R. Clark (@jasonrclark)

## Jellyfish 1.0.1 -- 2014-11-07

### Enhancements for Jellyfish core

* Now `Jellyfish.handle` could take multiple exceptions as the arguments.

### Other enhancements

* Introduced `Jellyfish::WebSocket` for basic websocket support.

## Jellyfish 1.0.0 -- 2014-03-17

### Incompatible changes

* Renamed `forward` to `cascade` to better aligned with Rack.

### Enhancements for Jellyfish core

* Introduced `log` and `log_error` for for controllers.
* Introduced `not_found` to trigger 404 response.
* Now we separate the idea of 404 and cascade. Use `not_found` for 404
  responses, and `cascade` for forwarding requests.

### Other enhancements

* Now we have Jellyfish::Swagger to generate Swagger documentation.
  Read README.md for more detail or checkout config.ru for a full example.

## Jellyfish 0.9.2 -- 2013-09-26

* Do not rescue Exception since we don't really want to rescue something
  like SignalException, which would break signal handling.

## Jellyfish 0.9.1 -- 2013-08-23

* Fixed a thread safety bug for initializing exception handlers.

## Jellyfish 0.9.0 -- 2013-07-11

### Enhancements for Jellyfish core

* We no longer use exceptions to control the flow. Use
  `halt(InternalError.new)` instead. However, raising exceptions
  would still work. Just prefer to use `halt` if you would like
  some performance boost.

### Incompatible changes

* If you're raising `NotFound` instead of using `forward` in your app,
  it would no longer forward the request but show a 404 page. Always
  use `forward` if you intend to forward the request.

## Jellyfish 0.8.0 -- 2013-06-15

### Incompatible changes

* Now there's no longer Jellyfish#controller but Jellyfish.controller,
  as there's no much point for making the controller per-instance.
  You do this to override the controller method instead:

``` ruby
class MyApp
  include Jellyfish
  def self.controller
    MyController
  end
  class MyController < Jellyfish::Controller
    def hi
      'hi'
    end
  end
  get{ hi }
end
```

* You can also change the controller by assigning it. The same as above:

``` ruby
class MyApp
  include Jellyfish
  class MyController < Jellyfish::Controller
    def hi
      'hi'
    end
  end
  controller MyController
  get{ hi }
end
```

### Enhancements for Jellyfish core

* Introduced Jellyfish.controller_include which makes it easy to pick
  modules to be included in built-in controller.
* Introduced Controller#halt as a short hand for `throw :halt`
* Now default route is `//`. Using `get{ 'Hello, World!' }` is effectively
  the same as `get(//){ 'Hello, World!' }`
* Now inheritance works.
* Now it raises TypeError if passing a route doesn't respond to :match.
* Now Jellyfish would find the most suitable error handler to handle
  errors, i.e. It would find the error handler which would handle the
  nearest exception class in the ancestors chain. Previously it would
  only find the first one which matches, ignoring the rest. It would
  also cache the result upon a lookup.

### Enhancements for Jellyfish extension

* Added `Jellyfish::ChunkedBody` which is similar to `Sinatra::Stream`.

* Added `Jellyfish::MultiAction` which gives you some kind of ability to do
  before or after filters. See README.md for usage.

* Added `Jellyfish::NormalizedParams` which gives you some kind of Sinatra

lib/jellyfish.rb
# frozen_string_literal: true

module Jellyfish
  autoload :VERSION , 'jellyfish/version'

  autoload :NormalizedParams, 'jellyfish/normalized_params'
  autoload :NormalizedPath  , 'jellyfish/normalized_path'

  autoload :Builder    , 'jellyfish/builder'
  autoload :Rewrite    , 'jellyfish/rewrite'
  autoload :ChunkedBody, 'jellyfish/chunked_body'
  autoload :WebSocket  , 'jellyfish/websocket'

  Cascade  = Object.new
  GetValue = Object.new

  class Response < RuntimeError
    def headers
      @headers ||= {'content-type' => 'text/html'}
    end

    def body
      @body ||= [File.read("#{Jellyfish.public_root}/#{status}.html")]
    end
  end

  class InternalError < Response; def status; 500; end; end
  class NotFound      < Response; def status; 404; end; end
  class Found         < Response # this would be raised in redirect
    attr_reader :url
    def initialize url; @url = url                             ; end
    def status        ; 302                                    ; end
    def headers       ; super.merge('location' => url)         ; end
    def body          ; super.map{ |b| b.gsub('VAR_URL', url) }; end
  end

  # -----------------------------------------------------------------

  class Controller
    attr_reader :routes, :jellyfish, :env
    def initialize routes, jellyfish
      @routes, @jellyfish = routes, jellyfish
      @status, @headers, @body = nil
    end

    def call env
      @env = env
      block_call(*dispatch)
    end

    def block_call argument, block
      val = instance_exec(argument, &block)
      [status || 200, headers || {}, body || rack_body(val || '')]
    rescue LocalJumpError
      log("Use `next' if you're trying to `return' or `break' from a block.")
      raise
    end

    def log     message; jellyfish.log(    message, env); end
    def log_error error; jellyfish.log_error(error, env); end
    def request   ; @request ||= Rack::Request.new(env); end
    def halt *args; throw(:halt, *args)                ; end
    def cascade   ;  halt(Jellyfish::Cascade)          ; end
    def not_found ;  halt(Jellyfish::NotFound.new)     ; end
    def found  url;  halt(Jellyfish::   Found.new(url)); end
    alias_method :redirect, :found

    def path_info     ; env['PATH_INFO']      || '/'  ; end
    def request_method; env['REQUEST_METHOD'] || 'GET'; end

    %w[status headers].each do |field|
      module_eval <<-RUBY
        def #{field} value=GetValue
          if value == GetValue
            @#{field}
          else
            @#{field} = value
          end
        end
      RUBY
    end

    def body value=GetValue
      if value == GetValue
        @body
      elsif value.nil?
        @body = value
      else
        @body = rack_body(value)
      end
    end

    def headers_merge value
      if headers.nil?
        headers(value)
      else
        headers(headers.merge(value))
      end
    end

    private
    def actions
      routes[request_method.downcase] || action_missing
    end

    def dispatch
      actions.find{ |(route, block)|
        case route
        when String
          break route, block if route == path_info
        else#Regexp, using else allows you to use custom matcher
          match = route.match(path_info)
          break match, block if match
        end
      } || action_missing
    end

    def action_missing
      if jellyfish.app then cascade else not_found end
    end

    def rack_body v
      if v.respond_to?(:each) || v.respond_to?(:to_path) then v else [v] end
    end
  end

  # -----------------------------------------------------------------

  module DSL
    def routes  ; @routes   ||= {}; end
    def handlers; @handlers ||= {}; end
    def handle *exceptions, &block
      exceptions.each{ |exp| handlers[exp] = block }
    end
    def handle_exceptions value=GetValue
      if value == GetValue
        @handle_exceptions
      else
        @handle_exceptions = value
      end
    end

    def controller_include *value
      (@controller_include ||= []).push(*value)
    end

    def controller value=GetValue
      if value == GetValue
        @controller ||= controller_inject(
          const_set(:Controller, Class.new(Controller)))
      else
        @controller   = controller_inject(value)
      end
    end

    def controller_inject value
      controller_include.
        inject(value){ |ctrl, mod| ctrl.__send__(:include, mod) }
    end

    %w[options get head post put delete patch].each do |method|
      define_method(method) do |route=//, meta={}, &block|
        define(method, route, meta, &block)
      end
    end

    def define(method, route=//, meta={}, &block)
      raise TypeError.new("Route #{route} should respond to :match") \
        unless route.respond_to?(:match)

      (routes[method.to_s] ||= []) << [route, block || lambda{|_|}, meta]
    end

    def inherited sub
      sub.handle_exceptions(handle_exceptions)
      sub.controller_include(*controller_include)
      [:handlers, :routes].each{ |m|
        val = __send__(m).inject({}){ |r, (k, v)| r[k] = v.dup; r }
        sub.__send__(m).replace(val) # dup the routing arrays
      }
    end
  end

  # -----------------------------------------------------------------

  def initialize app=nil; @app = app; end

  def call env
    ctrl = self.class.controller.new(self.class.routes, self)
    case res = catch(:halt){ ctrl.call(env) }
    when Cascade
      cascade(ctrl, env)
    when Response
      handle(ctrl, res, env)
    when Array
      res
    when NilClass # make sure we return rack triple
      ctrl.block_call(nil, lambda{|_|_})
    else
      raise TypeError.new("Expect the response to be a Jellyfish::Response" \
        " or Rack triple (Array), but got: #{res.inspect}")
    end
  rescue => error
    handle(ctrl, error, env)
  end

  def log_error error, env
    env['rack.errors']&.
      puts("[#{self.class.name}] #{error.inspect}" \
           " for #{env['PATH_INFO'] || '/'} #{error.backtrace}")
  end

  def log msg, env
    env['rack.errors']&.
      puts("[#{self.class.name}] #{msg}")
  end

  private
  def cascade ctrl, env
    app.call(env)
  rescue => error
    handle(ctrl, error, env)
  end

  def handle ctrl, error, env
    if handler = best_handler(error)
      ctrl.block_call(error, handler)
    elsif !self.class.handle_exceptions
      raise error
    elsif error.kind_of?(Response) # InternalError ends up here if no handlers
      [error.status, error.headers, error.body]
    else # fallback and see if there's any InternalError handler
      log_error(error, env)
      handle(ctrl, InternalError.new, env)
    end
  end

  def best_handler error
    handlers = self.class.handlers
    if handlers.key?(error.class)
      handlers[error.class]
    else # or find the nearest match and cache it
      ancestors = error.class.ancestors
      handlers[error.class] = handlers.dup. # thread safe iteration
        inject([nil, Float::INFINITY]){ |(handler, val), (klass, block)|
          idx = ancestors.index(klass) || Float::INFINITY # lower is better
          if idx < val then [block, idx] else [handler, val] end }.first
    end
  end

  # -----------------------------------------------------------------

  def self.included mod
    mod.__send__(:extend, DSL)
    mod.__send__(:attr_reader, :app)
    mod.handle_exceptions(true)
  end

  # -----------------------------------------------------------------

  module_function
  def public_root
    "#{File.dirname(__FILE__)}/jellyfish/public"
  end
end
